lighthouse-core/audits/seo/tap-targets.js

```diff
@@ -0,0 +1,274 @@
+const ViewportAudit = require('../viewport');
+const {
+  getRectOverlapArea,
+  getRectAtCenter,
+  allRectsContainedWithinEachOther,
+  getLargestRect,
+} = require('../../lib/rect-helpers');
+const {getTappableRectsFromClientRects} = require('../../lib/tappable-rects');
+
+const FINGER_SIZE_PX = 48;
+// Ratio of the finger area tapping on an unintended element
+// to the finger area tapping on the intended element
+const MAX_ACCEPTABLE_OVERLAP_SCORE_RATIO = 0.25;
+
+
+/**
+ * @param {LH.Artifacts.Rect} cr
+ */
+function clientRectMeetsMinimumSize(cr) {
+  return cr.width >= FINGER_SIZE_PX && cr.height >= FINGER_SIZE_PX;
+}
+
+/**
+ * @param {LH.Artifacts.TapTarget} target
+ */
+function targetIsTooSmall(target) {
+  for (const cr of target.clientRects) {
+    if (clientRectMeetsMinimumSize(cr)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} targets
+ */
+function getTooSmallTargets(targets) {
+  return targets.filter(targetIsTooSmall);
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget[]} tooSmallTargets
+ * @param {LH.Artifacts.TapTarget[]} allTargets
+ */
+function getOverlapFailures(tooSmallTargets, allTargets) {
+  /** @type {LH.Audit.TapTargetOverlapDetail[]} */
+  let failures = [];
+
+  tooSmallTargets.forEach(target => {
+    const overlappingTargets = getTooCloseTargets(
+      target,
+      allTargets
+    );
+
+    if (overlappingTargets.length > 0) {
+      failures = failures.concat(overlappingTargets);
+    }
+  });
+
+  return failures;
+}
+
+/**
+ *
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget[]} allTapTargets
+ */
+function getTooCloseTargets(tapTarget, allTapTargets) {
+  /** @type LH.Audit.TapTargetOverlapDetail[] */
+  const failures = [];
+
+  for (const maybeOverlappingTarget of allTapTargets) {
+    if (maybeOverlappingTarget === tapTarget) {
+      // checking the same target with itself, skip
+      continue;
+    }
+
+    const failure = getTargetTooCloseFailure(tapTarget, maybeOverlappingTarget);
+    if (failure) {
+      failures.push(failure);
+    }
+  }
+
+  return failures;
+}
+
+/**
+ * @param {LH.Artifacts.TapTarget} tapTarget
+ * @param {LH.Artifacts.TapTarget} maybeOverlappingTarget
+ * @returns {LH.Audit.TapTargetOverlapDetail | null}
+ */
+function getTargetTooCloseFailure(tapTarget, maybeOverlappingTarget) {
+  // convert client rects to unique tappable areas from a user's perspective
+  const tappableRects = getTappableRectsFromClientRects(tapTarget.clientRects);
+  const isHttpOrHttpsLink = /https?:\/\//.test(tapTarget.href);
+  if (isHttpOrHttpsLink && tapTarget.href === maybeOverlappingTarget.href) {
+    // no overlap because same target action
+    return null;
+  }
+
+  /** @type LH.Audit.TapTargetOverlapDetail | null */
+  let greatestFailure = null;
+  for (const targetCR of tappableRects) {
+    if (allRectsContainedWithinEachOther(tappableRects, maybeOverlappingTarget.clientRects)) {
```

Comment:
seems like this check can be run just once outside the loop (doesn't use `targetCR` at all)

is this even necessary or will running through the loop yield the same result?

Reply:
Good catch, we don't need it anymore because we now filter out child tap targets in the gatherer.